protocol/bin/flow.js

```diff
@@ -428,24 +390,41 @@ export type Long = number
 export type String = string
 export type Uint = number
 export type Uint64 = number
+type IncomingErrorCallback = (?{code?: number, desc?: string}) => void
+type IncomingReturn = Effect | null | void | false | Array<Effect | null | void | false>
 `
   const consts = Object.keys(typeDefs.consts).map(k => typeDefs.consts[k])
   const types = Object.keys(typeDefs.types).map(k => typeDefs.types[k])
-  const messageResponse = Object.keys(typeDefs.messages).map(k => typeDefs.messages[k].response)
-  const messageParams = Object.keys(typeDefs.messages).map(k => typeDefs.messages[k].paramType)
   const messagePromise = Object.keys(typeDefs.messages).map(k => typeDefs.messages[k].rpcPromiseType)
   const messageChannelMap = Object.keys(typeDefs.messages).map(k => typeDefs.messages[k].rpcChannelMapType)
   const messageEngineSaga = Object.keys(typeDefs.messages).map(k => typeDefs.messages[k].engineSagaType)
   const callMapType = Object.keys(project.incomingMaps).length ? 'IncomingCallMapType' : 'void'
   const incomingMap =
-    `\nexport type IncomingCallMapType = {|` +
+    `\nexport type IncomingCallMapType<TypedState> = {|` +
     Object.keys(project.incomingMaps)
-      .map(im => `  '${im}'?: ${project.incomingMaps[im]}`)
+      .map(im => `  ${im}?: ${project.incomingMaps[im]}`)
       .join(',') +
     '|}'
+
+  const messageTypesData = Object.keys(typeDefs.messages)
+    .map(k => {
+      const data = typeDefs.messages[k]
+      const types = {}
+      return `  ${k}: {|
+    inParam: ${data.inParam},
+    outParam: ${data.outParam || 'void'},
+  |},`
+    })
+    .sort()
+    .join('\n')
+
+  const messageTypes = `\nexport type MessageTypes = {|
+${messageTypesData}
+|}`
+
   const data = [
-    ...[...consts, ...types, ...messageResponse].sort(),
-    ...messageParams.sort(),
+    messageTypes,
```

Comment:
best to look at the diff of the generated to get the gist of the changes here. We basically make a map of method => metaData, then the helpers now just pull that metadata using $PropertyType<CallMap, 'inParams'> to extract the type info. This makes the helpers not actually reference any of the types directly at all and its very clear what all the related data is in the map on top